src/order/concept.lean

```diff
@@ -0,0 +1,258 @@
+(gc_intent_closure_extent_closure r).l_supr₂
+
+@[simp] lemma extent_closure_Union₂ (f : Π i, κ i → set β) :
+  extent_closure r (⋃ i j, f i j) = ⋂ i j, extent_closure r (f i j) :=
+intent_closure_Union₂ _ _
+
+lemma subset_extent_closure_intent_closure (s : set α) :
+  s ⊆ extent_closure r (intent_closure r s) :=
+(gc_intent_closure_extent_closure r).le_u_l _
+
+lemma subset_intent_closure_extent_closure (t : set β) :
+  t ⊆ intent_closure r (extent_closure r t) :=
+subset_extent_closure_intent_closure _ t
+
+@[simp] lemma intent_closure_extent_closure_intent_closure (s : set α) :
+  intent_closure r (extent_closure r $ intent_closure r s) = intent_closure r s :=
+(gc_intent_closure_extent_closure r).l_u_l_eq_l _
+
+@[simp] lemma extent_closure_intent_closure_extent_closure (t : set β) :
+  extent_closure r (intent_closure r $ extent_closure r t) = extent_closure r t :=
+intent_closure_extent_closure_intent_closure _ t
+
+lemma intent_closure_anti : antitone (intent_closure r) :=
+(gc_intent_closure_extent_closure r).monotone_l
+
+lemma extent_closure_anti : antitone (extent_closure r) := intent_closure_anti _
+
+/-! ### Concepts -/
+
+variables (α β)
+
+/-- The formal concepts of a relation. A concept of `r : α → β → Prop` is a pair of sets `s`, `t`
+such that `s` is the set of all elements that are `r`-related to all of `t` and `t` is the set of
+all elements that are `r`-related to all of `s`. -/
+structure concept extends set α × set β :=
+(closure_fst : intent_closure r fst = snd)
+(closure_snd : extent_closure r snd = fst)
+
+namespace concept
+variables {r α β} {c d : concept α β r}
+
+attribute [simp] closure_fst closure_snd
+
+@[ext] lemma ext (h : c.fst = d.fst) : c = d :=
+begin
+  obtain ⟨⟨s₁, t₁⟩, h₁, _⟩ := c,
+  obtain ⟨⟨s₂, t₂⟩, h₂, _⟩ := d,
+  dsimp at h₁ h₂ h,
+  subst h,
+  subst h₁,
+  subst h₂,
+end
+
+lemma ext' (h : c.snd = d.snd) : c = d :=
+begin
+  obtain ⟨⟨s₁, t₁⟩, _, h₁⟩ := c,
+  obtain ⟨⟨s₂, t₂⟩, _, h₂⟩ := d,
+  dsimp at h₁ h₂ h,
+  subst h,
+  subst h₁,
+  subst h₂,
+end
+
+lemma fst_injective : injective (λ c : concept α β r, c.fst) := λ c d, ext
+lemma snd_injective : injective (λ c : concept α β r, c.snd) := λ c d, ext'
+
+instance : has_sup (concept α β r) :=
+⟨λ c d, { fst := extent_closure r (c.snd ∩ d.snd),
+  snd := c.snd ∩ d.snd,
+  closure_fst := by rw [←c.closure_fst, ←d.closure_fst, ←intent_closure_union,
+    intent_closure_extent_closure_intent_closure],
+  closure_snd := rfl }⟩
+
+instance : has_inf (concept α β r) :=
+⟨λ c d, { fst := c.fst ∩ d.fst,
+  snd := intent_closure r (c.fst ∩ d.fst),
+  closure_fst := rfl,
+  closure_snd := by rw [←c.closure_snd, ←d.closure_snd, ←extent_closure_union,
+    extent_closure_intent_closure_extent_closure] }⟩
+
+instance : semilattice_inf (concept α β r) := fst_injective.semilattice_inf _ $ λ _ _, rfl
+
+@[simp] lemma fst_subset_fst_iff : c.fst ⊆ d.fst ↔ c ≤ d := iff.rfl
+@[simp] lemma fst_ssubset_fst_iff : c.fst ⊂ d.fst ↔ c < d := iff.rfl
+
```

Comment:
Could you add a pair of lemmas stating  `strict_mono fst` and `strict_anti snd`?